### ISSUE TITLE
Got F811 and F821 when indirectly importing Literal and overload

### ISSUE THREAD

**Issue by leavers** [astral-sh/ruff#6969]
A minimal code snippet that reproduces the bug:

File `test.py`

``` python
from mytyping import Literal, overload
# from typing import Literal, overload
# from typing_extensions import Literal, overload

tag: Literal["a"] = "a"


@overload
def func(a: int):
    ...


@overload
def func(a: int, b: int):
    ...


def func(a: int, b: int = 0):
    return a, b
```

File `mytyping.py`:

``` python
from typing import Literal, overload

__all__ = ["Literal", "overload"]
```

When using `ruff check` with `from mytyping import Literal, overload` enabled, I got:

``` shell
$ ruff check test.py
test.py:5:15: F821 Undefined name `a`
test.py:14:5: F811 Redefinition of unused `func` from line 9
test.py:18:5: F811 Redefinition of unused `func` from line 14
Found 3 errors.
```

However, when uncomment either `typing` or `typing_extensions`, `ruff check` does not report above errors.

Ruff version:

``` shell
$ ruff --version
ruff 0.0.286
```

**Comment by leavers** [comment]
@charliermarsh Thanks for your reply, this setting resolved my issue!

**Comment by rj-xu** [comment]
```python
# from typing import overload
from my_typing import overload

@overload
def func(a: str, b: int) -> int: ...
@overload
def func(a: int, b: str) -> int: ...
def func(a: int | str, b: int | str) -> int:
    return 0
```
In version 0.12.5, `F811 Redefinition` still occurs, even though `typing-modules` have already been set up.
This error will appear after v0.3.0.

**Comment by MichaReiser** [comment]
@rj-xu I do appreciate that you took the time to find this issue but please open a new issue if you need help. It will make it easier for us to keep track where users still need help with (which is hard if people comment on issues that have been closed for a long time). 

In the new issue, can you link to this issue for context, include a code snippet and your configuration? This should allow us to help you. Thank you
